### PR TITLE
fix(executor): Fix resource patch when not providing flags. Fixes #5310

### LINKED ISSUE #5310: Unable to patch a resource
Hello,

I am trying to patch `sa` to add an `imagePullSecret`

Here is the template I am using:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: ClusterWorkflowTemplate
metadata:
  name: add-image-pull-secret
spec:
  entrypoint: execute
  podGC:
    strategy: OnPodSuccess
  retryStrategy:
   retryPolicy: OnError
   limit: 1
  templates:
    - name: execute
      inputs: 
        parameters: [
          {name: "namespace"},
          {name: "secret", value: "github-container-registry"}
        ]
      resource:
        action: patch
        manifest: |
          apiVersion: v1
          kind: ServiceAccount
          metadata:
            name: default
            namespace: {{inputs.parameters.namespace}}
          imagePullSecrets:
          - name: {{inputs.parameters.secret}}
  tolerations:
    - effect: NoSchedule
      key: kubernetes.azure.com/scalesetpriority
      operator: Equal
      value: spot
```

When I run this, I keep getting this error:
```
executor error: Must provide at least one of flags or manifest.
```

Here is the full log from the step:
```
time="2021-03-05T14:41:14.433Z" level=info msg="Starting Workflow Executor" version="{latest+ae5587e.dirty 2021-03-02T21:48:50Z ae5587e97dad0e4806f7a230672b9
98fe140a767 v3.0.0-rc4 dirty go1.15.7 gc linux/amd64}"
time="2021-03-05T14:41:14.460Z" level=info msg="Creating a docker executor"
time="2021-03-05T14:41:14.460Z" level=info msg="Executor (version: latest+ae5587e.dirty, build_date: 2021-03-02T21:48:50Z) initialized (pod: ky-cloud-wor
kflows/provision-wp-st-mp-b6c56aa98d6a-1998521980) with template:\n{\"name\":\"execute\",\"inputs\":{\"parameters\":[{\"name\":\"namespace\",\"v
alue\":\"kcs-mp\"},{\"name\":\"secret\",\"value\":\"github-container-registry\"}]},\"outputs\":{},\"metadata\":{},\"resource\":{\"action\":\"patch\",\"me
rgeStrategy\":\"strategic\",\"manifest\":\"apiVersion: v1\\nkind: ServiceAccount\\nmetadata:\\n  name: default\\n  namespace: kcs-mp\\nimagePullSecrets:\
\n- name: github-container-registry\\n\"}}"
time="2021-03-05T14:41:14.460Z" level=info msg="Loading manifest to /tmp/manifest.yaml"
time="2021-03-05T14:41:14.461Z" level=error msg="executor error: Must provide at least one of flags or manifest.\ngithub.com/argoproj/argo-workflows/v3/error
s.New\n\t/go/src/github.com/argoproj/argo-workflows/errors/errors.go:49\ngithub.com/argoproj/argo-workflows/v3/workflow/executor.(*WorkflowExecutor).getKubec
tlArguments\n\t/go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:101\ngithub.com/argoproj/argo-workflows/v3/workflow/executor.(*Workfl
owExecutor).ExecResource\n\t/go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:31\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/c
ommands.execResource\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:48\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/c
ommands.NewResourceCommand.func1\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:25\ngithub.com/spf13/cobra.(*Command).execut
e\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/comman
d.go:950\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887\nmain.main\n\t/go/src/github.com/argoproj/arg
o-workflows/cmd/argoexec/main.go:14\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:204\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1374"
time="2021-03-05T14:41:14.461Z" level=fatal msg="Must provide at least one of flags or manifest.\ngithub.com/argoproj/argo-workflows/v3/errors.New\n\t/go/src
/github.com/argoproj/argo-workflows/errors/errors.go:49\ngithub.com/argoproj/argo-workflows/v3/workflow/executor.(*WorkflowExecutor).getKubectlArguments\n\t/
go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:101\ngithub.com/argoproj/argo-workflows/v3/workflow/executor.(*WorkflowExecutor).Exec
Resource\n\t/go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:31\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.execReso
urce\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:48\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.NewResou
rceCommand.func1\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:25\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod
/github.com/spf13/cobra@v1.0.0/command.go:846\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950\ngithub
.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887\nmain.main\n\t/go/src/github.com/argoproj/argo-workflows/cmd/
argoexec/main.go:14\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:204\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1374"
```

### PR DESCRIPTION
**Pull request by terrytangyuan**
Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

**Comment by terrytangyuan** [comment]
Seems like a bug introduced in https://github.com/argoproj/argo-workflows/pull/4908 and a regression since v3.0.0-rc1.

File: workflow/executor/resource.go [hunk 1]
```diff
@@ -97,7 +97,8 @@ func (we *WorkflowExecutor) getKubectlArguments(action string, manifestPath stri
 	if len(buff) != 0 && action != "patch" {
 		args = append(args, "-f")
 		args = append(args, manifestPath)
-	} else if len(flags) <= 0 {
+	}
+	if len(flags) <= 0 && len(buff) == 0 {
```
Comment (simster7) [inline workflow/executor/resource.go:101]: Minor: Why would `len(flags)` return anything less than 0?
```suggestion
	if len(flags) == 0 && len(buff) == 0 {
```

Comment (terrytangyuan) [reply workflow/executor/resource.go]: Thanks! Updated.

File: workflow/executor/resource_test.go [hunk 2]
```diff
@@ -34,8 +37,51 @@ func TestResourceFlags(t *testing.T) {
...
 		RuntimeExecutor:    &mockRuntimeExecutor,
 	}
-	args, err := we.getKubectlArguments("fake", "../../examples/hello-world.yaml", fakeFlags)
-
+	args, err := we.getKubectlArguments("fake", manifestPath, fakeFlags)
 	assert.Nil(t, err)
 	assert.Contains(t, args, fakeFlags[0])
+
+	_, err = we.getKubectlArguments("fake", manifestPath, nil)
+	assert.Nil(t, err)
```
Comment (alexec) [inline workflow/executor/resource_test.go:45]: minor - prefer NoError

**Review by alexec — COMMENTED** [review]
@book987 can you please take a look?

File: workflow/executor/resource.go [hunk 3]
```diff
@@ -97,7 +97,8 @@ func (we *WorkflowExecutor) getKubectlArguments(action string, manifestPath stri
 	if len(buff) != 0 && action != "patch" {
 		args = append(args, "-f")
 		args = append(args, manifestPath)
-	} else if len(flags) <= 0 {
+	}
+	if len(flags) == 0 && len(buff) == 0 {
```
Comment (book987) [inline workflow/executor/resource.go:101]: Minor: should this block be placed at the top of this func before collecting args, as another validation of user inputs?

**Comment by terrytangyuan** [comment]
Rebased and finally passed the CI. @alexec Thanks for making the CI more robust!

**Comment by wreed4** [comment]
@alexec it seems like this fix is still not available in 3.0.1, 3.0.2, or 3.0.3.  I can see the cherry-picks above, but I can't see it in the changelogs, and most importantly, we have verified the bug still exists in all three of those versions.  believe we tested 3.0.4 as well but not positive.  Can you give any insight into when this will be available?

**Comment by wreed4** [comment]
upon further looking actually.... looks like this particular code is there, but the syntax for patch must have changed from 2.X?? Are there docs on that I can't seem to find them.  I'm now getting this error: `Error (exit code 1): error: You must provide one or more resources by argument or filename.` with this patch
```yaml
  - name: patch
    resource:
      action: patch
      manifest: |
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: patch-test-{{workflow.name}}
          annotations:
            patched: "true"
```

the same code works in 2.x

**Comment by wreed4** [comment]
It's this line here https://github.com/argoproj/argo-workflows/blob/master/workflow/executor/resource.go#L107
It prevents passing the file so kubectl doesn't know which object to pass.  I'd definitely consider this a regression. Docs seems like they are not updated either to reflect an intentional change.  https://github.com/argoproj/argo-workflows/tree/v3.0.4/examples#kubernetes-resources

**Pull request merged**